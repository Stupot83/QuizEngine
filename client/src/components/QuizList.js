import React, { Component } from "react";
import { connect } from "react-redux";
import Modal from "./Modal";
import PropTypes from "prop-types";
import Grid from "@material-ui/core/Grid";
import Card from "@material-ui/core/Card";
import Typography from "@material-ui/core/Typography";
import Button from "@material-ui/core/Button";
import AddCircleOutlineIcon from "@material-ui/icons/AddCircleOutline";
import EditIcon from "@material-ui/icons/Edit";
import VisibilityIcon from "@material-ui/icons/Visibility";
import Logo from "../Logo-no-quizzes.png";
import "../sass/QuizList.scss";

class QuizList extends Component {
    state = {
        modal: false,
        edit: false,
        title: "",
        category: "",
        id: "",
        creator: {}
    };

    toggleModal = e => {
        this.setState({ modal: !this.state.modal, edit: false });
    };

    toggleEditModal = (title, category, id, creator, e) => {
        e.stopPropagation();

        this.setState({
            modal: !this.state.modal,
            edit: !this.state.edit,
            title: title,
            category: category,
            id: id,
            creator: creator
        });
    };

    render() {
        const { quizzes } = this.props.quizzes;
        const userRole = this.props.auth.user.role;

        let content;

        let quizData = quizzes.sort().map(quiz => (
            <Card key={quiz._id} className="Quiz_card">
                <Grid item xs={12} className="Quiz_header_title">
                    Title:
                    <Typography id="actualQuizTitle" className="Quiz_title">
                        {quiz.title}
                    </Typography>
                </Grid>
                <Grid item xs={12} className="Quiz_header_category">
                    Category:
                    <Typography id="actualQuizCategory" className="Quiz_category">
                        {quiz.category}
                    </Typography>
                </Grid>
                {userRole === "Edit" && (
                    <Grid item xs={12}>
                        <Button
                            id="editQuizButton"
                            variant="contained"
                            className="Quiz_button_edit"
                            size="large"
                            startIcon={<EditIcon />}
                            onClick={this.toggleEditModal.bind(this, quiz.title, quiz.category, quiz._id, quiz.creator)}
                        >
                            Edit Quiz
                        </Button>
                    </Grid>
                )}
                <Grid item xs={12}>
                    <Button
                        id="viewQuizButton"
                        variant="contained"
                        className="Quiz_button_view"
                        startIcon={<VisibilityIcon />}
                        size="large"
                        onClick={() => this.props.history.push(`/quizzes/${quiz._id}`)}
                    >
                        View Quiz
                    </Button>
                </Grid>
            </Card>
        ));

        if (quizzes.length > 0) {
            // When there is one or more Quizzes
            content = (
                <>
                    {userRole === "Edit" && (
                        <Grid item xs={12} className="Create_quiz_button_container">
                            <Button
                                variant="contained"
                                size="large"
                                className="Create_quiz_button"
                                startIcon={<AddCircleOutlineIcon />}
                                onClick={this.toggleModal}
                                id="createQuizButton"
                            >
                                Create Quiz
                            </Button>
                        </Grid>
                    )}
                    <div className="Modal_area">
                        <Modal
                            onClose={this.toggleModal}
                            modal={this.state.modal}
                            edit={this.state.edit}
                            title={this.state.title}
                            category={this.state.category}
                            id={this.state.id}
                            creator={this.state.creator}
                        />
                    </div>
                    <Grid container alignItems="center" justify="center" className="Quizzes_container">
                        {quizData}
                    </Grid>
                </>
            );
        } else {
            // No Quizzes
            content = (
                <>
                    {userRole === "Edit" && (
                        <Grid container alignItems="center" justify="center" className="No_quizzes_container">
                            <Grid item xs={12}>
                                <img id="noQuizzesLogo" className="Logo_no_quizzes" src={Logo} alt="Loading" />
                            </Grid>
                            <Grid item xs={12}>
                                <Button
                                    className="Create_quiz_button"
                                    variant="contained"
                                    size="large"
                                    startIcon={<AddCircleOutlineIcon />}
                                    onClick={this.toggleModal}
                                    id="createQuizButton"
                                >
                                    Create Quiz
                                </Button>
                            </Grid>
                            <div className="Modal_area">
                                <Modal onClose={this.toggleModal} modal={this.state.modal} />
                            </div>
                        </Grid>
                    )}
                </>
            );
        }

        return (
            <Card className="QuizList_container">
                <h1 id="header" className="QuizList_header">
                    QuizList
                </h1>
                {content}
            </Card>
        );
    }
}

QuizList.propTypes = {
    auth: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
    auth: state.auth,
    quizzes: state.quizzes
});

export default connect(mapStateToProps, {})(QuizList);
