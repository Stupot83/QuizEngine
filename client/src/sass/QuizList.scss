.QuizList_container {
    flex: 1 1 auto;
    margin: 4rem;
    border: 1px solid #efefef;
    border-radius: 30px !important;
    background-image: url("../corkboard.png");
    background-size: 100% 100%;

    .QuizList_header {
        font-size: 24px;
        color: ghostwhite;
        text-align: center;
        padding: 40px;
        background-color: transparent;
        text-decoration: underline;
    }

    .Create_quiz_button_container {
        text-align: center;
        background-color: transparent;

        .Create_quiz_button {
            min-width: 150px;
            margin-left: 30px;
            margin-right: 30px;
            padding: 15px;
            border-radius: 15px;
            background-color: #34d1bf;
            color: #070707;
            font-weight: bold;
        }
    }

    .Modal_area {
        display: flex;
        justify-content: center;
    }

    .Quizzes_container {
        padding: 30px;
        display: flex;
        flex-wrap: wrap;
        background-color: transparent;

        .Quiz_card {
            height: 400px;
            min-width: 350px;
            flex-basis: 300px;
            border-radius: 30px;
            display: flex;
            flex-direction: column;
            align-items: center;
            border: 1px solid #efefef;
            margin: 3rem;
            text-align: center;
            color: #070707;
            background-color: ghostwhite !important;

            .Quiz_header_title {
                margin-top: 20px;
            }

            .Quiz_header_category {
                margin-top: 0px;
                margin-bottom: 10px;
            }

            .Quiz_title {
                text-transform: uppercase;
                color: #070707;
                font-size: 18px;
                height: 10vh;
                font-weight: bold;
                height: 100%;
                padding: 20px !important;
                white-space: nowrap;
                overflow: hidden;
                display: block;
                text-overflow: ellipsis;
            }

            .Quiz_category {
                text-transform: uppercase;
                color: #070707;
                font-size: 18px;
                height: 10vh;
                font-weight: bold;
                height: 100%;
                padding: 20px !important;
                white-space: nowrap;
                overflow: hidden;
                display: block;
                text-overflow: ellipsis;
            }

            .Quiz_button_edit {
                min-width: 150px;
                margin-left: 30px;
                margin-right: 30px;
                padding: 15px;
                border-radius: 15px;
                background-color: #3454d1;
                color: ghostwhite;
                font-weight: bold;
            }

            .Quiz_button_view {
                min-width: 150px;
                margin-left: 30px;
                margin-right: 30px;
                padding: 15px;
                border-radius: 15px;
                background-color: #34d1bf;
                color: #070707;
                font-weight: bold;
            }
        }
    }

    .No_quizzes_container {
        padding-bottom: 50px;
        display: flex;
        flex-wrap: wrap;
        background-color: transparent;
        flex-direction: column;

        .Create_quiz_button {
            min-width: 150px;
            margin-left: 30px;
            margin-right: 30px;
            padding: 15px;
            border-radius: 15px;
            background-color: #34d1bf;
            color: #070707;
            font-weight: bold;
        }

        .Logo_no_quizzes {
            margin: 40px;
            border-radius: 20%;
            border: 10px solid #efefef;
            height: 400px;
            width: 400px;
        }
    }
}
