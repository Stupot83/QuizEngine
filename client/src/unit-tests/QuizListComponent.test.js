import React from "react";
import { shallow } from "enzyme";
import QuizList from "../components/QuizList";
import QuizEngineStore from "../quizEngineStore";
import { Provider } from "react-redux";

const mockStore = QuizEngineStore;

describe("QuizList component", () => {
    it("should render correctly in debug mode", () => {
        const component = shallow(
            <Provider store={mockStore}>
                <QuizList debug />
            </Provider>
        );

        expect(component).toMatchSnapshot();
    });

    it("should render correctly", () => {
        const component = shallow(
            <Provider store={mockStore}>
                <QuizList />
            </Provider>
        );

        expect(component).toMatchSnapshot();
    });
});
