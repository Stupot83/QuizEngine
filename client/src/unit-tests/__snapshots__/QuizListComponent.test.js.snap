// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizList component should render correctly 1`] = `ShallowWrapper {}`;

exports[`QuizList component should render correctly in debug mode 1`] = `ShallowWrapper {}`;
